### ISSUE TITLE
Failure error when used as a shell script with Hazel

### ISSUE THREAD

**Issue by rEes9P** [qpdf/qpdf#907]
I've been using this excellent utility with the equally impressive Hazel app on my Mac to decrypt password protected account statements and then have Hazel file them. Of late, I've been getting failure error notifications from Hazel. Despite these errors, the script seems to be working and doing its job. Here is the simple script that I have running in Hazel (using shell /bin/zsh):
`qpdf --decrypt --password=somepassword --replace-input "$1"`

And here is a typical output of an error:
```
2023-02-20 15:09:04.651 hazelworker[14558] [Error] Shell script failed: Error processing shell script on file /Users/johndoe/Documents/Statements/Card/2023/02-20-2023.pdf.
2023-02-20 15:09:04.651 hazelworker[14558] Shellscript exited with non-successful status code: 3
```

As a result of the error message the process stops and the subsequent rules in the Hazel rule sequence do not get to run. Would love some help in getting to the bottom of this. Thanks!

**Comment by jberkenbilt** [comment]
That means qpdf is finding and issuing warnings on the file. You can run the command from the command line and see what the warnings are. If you are confident that qpdf is not damaging the files and are willing to ignore warnings, you can pass `--no-warn` and `--warning-exit-0` to qpdf, and it will then silently ignore warnings and not exit with code 3. For details, see [Exit Status in the docs](https://qpdf.readthedocs.io/en/stable/cli.html#exit-status). I hope this helps.

**Comment by rEes9P** [comment]
Thanks for this, @jberkenbilt. I ran it from the command line and get the following warning:
```
WARNING: /Users/johndoe/Documents/Statements/Card/2023/02-20-2023.pdf (encryption dictionary, offset 558): /Perms field in encryption dictionary doesn't match expected value
qpdf: there are warnings; original file kept in /Users/johndoe/Documents/Statements/Card/2023/02-20-2023.pdf.~qpdf-orig
qpdf: operation succeeded with warnings; resulting file may have some problems
```

When I ran the command with the two conditions (`--no-warn` and `--warning-exit-0`) I get the following output:
`qpdf: there are warnings; original file kept in /Users/johndoe/Documents/Statements/Card/2023/02-20-2023.pdf.~qpdf-orig`

I expected the warning to not be there after the two additional commands, nor for the original to be kept alongside. Is the command I have used correct in the way it is sequenced?
`qpdf --decrypt --password=example --replace-input filename.pdf --no-warn --warning-exit-0`

Thanks again for your help as always!

**Comment by jberkenbilt** [comment]
Yeah, I forgot about the interaction between `--replace-input` and warnings. Basically if qpdf had warnings, it conservatively refuses to throw away the original. The last thing you want is for qpdf to have destroyed your file and the original to be gone. If you have the original in a safe place in case qpdf does damage, you can always bypass `--replace-input` and just write to a temporary file and rename over top of the original. qpdf has been around for over 20 years and has been open source for 15, and it has processed no doubt billions of files. My employer alone processes over a million pages a day with it, and we're a small company. It basically never damages a correct PDF file, and it seldom damages a file with warnings, but I have seen cases where qpdf loses input data or causes some damage in files that it generates warnings for. If this is something you're doing with files that are generated in a consistent fashion and you are confident that qpdf is not hurting the file, you can throw away the original, but I'm not likely to add a feature to let qpdf throw away the original with `--replace-input` if there were warnings. In other words, qpdf is extremely reliable, but it is not perfect. :-)

I can tell you that the warning above is innocuous and that it would be safe to ignore it. A long-requested feature is for qpdf's warnings to be put into various categories, such as warnings that matter to a user (like qpdf attempting to recover from a broken file), things that matter to a general developer (like the warning above that probably points to a bug in the code that created the PDF), and things that matter to someone writing software that uses qpdf that a user couldn't do anything about, like object type warnings that may reflect incorrect assumptions on the part of the developer about PDF file structure. (Those warnings have prevented me from releasing broken code more than once and have occasionally enabled someone else to report a bug.)

**Comment by rEes9P** [comment]
Thanks @jberkenbilt. Yes, the PDFs in question seem to be just fine so I will go ahead and ignore the errors.

>My employer alone processes over a million pages a day with it
>

Trying to wrap my head around that number!

**Comment by rubyFeedback** [comment]
> Trying to wrap my head around that number!

That probably can happen if you run a business and need to process tons of .pdf files.

I recently had a new use case. An elderly relative has memory-related problems, and is decreasingly able to work with computers, printing out .pdf files (well ...). Trying to explain the issue with that led to no change, so I realised I will need to do more automatic processing of .pdf files in the future, as that relative constantly phones me to then fix issues that even young, non-tech savvy people can easily solve these days.

**Comment by jberkenbilt** [comment]
@rubyFeedback I can sympathize with this situation. qpdf is not a world-changing application, but it's nice to know when it people can actually use it to help people. Thanks for sharing.

Re millions of pages, my employer processes medical records, sometimes for large companies that process insurance claims. It's not uncommon for medical records to have hundreds or thousands, or in some cases, even tens of thousands of pages. We pull in the complete records and split into individual pages for storage, then sometimes recombine the pages. So, yes, millions of pages a day run through the system. When I wrote qpdf, I worked for a data conversion company. We processed PDF files on a large scale there too. The volume was only about 1% of my current employer, but there was a lot more variety in the files. I have also had conversations with other people who use qpdf as part of large document automation workflows, so qpdf has seen a vast number of files in its life.
